### DISCUSSION TITLE
인덱스는 크게 Hash 인덱스와 B+Tree 인덱스가 있습니다. 이것은 무엇일까요?

### DISCUSSION THREAD

**Discussion by eatingbug** [Team-Psyduck/tech-sharing#68] [General]
- 오늘의 알고리즘 : [DP - 11727](https://www.acmicpc.net/problem/11727)

**Comment by eatingbug** [comment]
- B-Tree 인덱스는 이진 탐색 트리의 일반화된 형태로, 메모리에 담기 어려운 큰 데이터를 다루기위해 사용된다. 항상 정렬된 상태를 유지한다.
  - 파일시스템이나 데이터베이스에 적합하다.
  - 노드의 개수를 늘리고 트리의 전체 높이를 줄여서 탐색속도를 높힐 수 있다.
- B+Tree 는 B-Tree 를 개선시킨 자료구조다. B-Tree와 다르게 오직 리프 노드의 키만 데이터 포인트를 갖고 있다. 그래서 더 많은 키를 보관할 수 있다.
  - 많은 DBMS 에서 B-Tree 대신 B+Tree 를 사용한다.
- 해시 인덱스는 Key-Value 쌍으로 데이터를 저장하는 인덱스 방식이다. Key 값을 알고있다면 시간복잡도가 O(1) 지만, 데이터가 정렬되어 있지 않아서 부등호 연산에 부적합하다.

출처 : https://hudi.blog/db-index-and-indexing-algorithms/

**Comment by SionBackEnd** [comment]
Hash인덱스란 Hash값을 이용하여 key와 value 형태로 저장하는 인덱싱 방식이다. Tree 방식보다 조회가 더 빠르다는 장점을 갖고 있지만, 범위 조회를 할 때 Tree방식보다 느리다는 단점이 되어있다. 

B+Tree는 맨 하단 리프를 제외하고 포인터들로 이루어진 인덱싱 방식이다. 포인터에는 특정 값들이 저장되어 있어서 해당 포인터를 통해 특정 값을 갖고있는 데이터를 빠르게 찾을 수 있다. Hash인덱스보다 조회속도는 느리지만, 범위 조회시 첫 번째 값을 찾아낸 후 오른쪽 방향으로 이동하기만 하면 되기 떄문에 가장 많이 사용되는 인덱싱 자료구조이다.
